### PR TITLE
update README to include latest APIs

### PR DESCRIPTION
**Pull request by exdx**
Signed-off-by: Daniel Sover <dsover@redhat.com>

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -13,3 +13,6 @@
 
 # Dependency directories
 vendor/
+
+# goland 
+.idea/*
```
Comment (njhale) [inline .gitignore:18]: I'd like to keep the `.gitignore` sparse if we can -- @exdx maybe it's better for this to go in global git settings per contributor?

File: README.md [hunk 2]
```diff
@@ -1,3 +1,102 @@
...
+`Bundles` are used to reference content that may need to be pulled and should be unpacked. 
+In this sense, Bundle is a generalization of the image concept, and can be used to represent any type of content.
+
+`Bundles` do nothing on their own - they require a `Provisioner` to unpack and make their content available in-cluster. They can be
+unpacked to any arbitrary storage medium such as a PersistentVolume or ConfigMap. Each `Bundle` has an associated `spec.class` field which 
+indicates the `Provisioner` that should be watching and unpacking that particular bundle type. 
+
+Example Bundle
+```yaml
+apiVersion: core.RukPak.io/v1alpha1
```
Comment (njhale) [inline README.md:30]: nit: `core.rukpak.io/v1alpha1`

Comment (exdx) [reply README.md:30]: whoops, find and replace did this

File: README.md [hunk 3]
```diff
@@ -1,3 +1,102 @@
...
+RukPak is a pluggable solution for the packaging and distribution of cloud-native content and supports advanced strategies for installation, updates, and policy.
+The project provides a content ecosystem for installing a variety of artifacts, such as Git repositories, Helm charts, OLM bundles, and more onto a Kubernetes cluster. 
+These artifacts can then be managed, scaled, and upgraded in a safe way to enable powerful cluster extensions. 
+
+At its core, RukPak is a small set of APIs, packaged as Kubernetes CustomResourceDefinitions, and controllers that watch for those APIs. These APIs express what 
+content is being installed on-cluster and how to create a running instance of the content. 
+
+## Components 
+
+RukPak is composed of a few primary APIs: `Bundle`, `Instance`, and `Provisioner/ProvisionerClass`
```
Comment (timflannagan) [inline README.md:16]: I'd remove the mention of Provisioner here - it seems to imply that it's a concrete type vs. the controller responsible for watching these primary APIs.

File: README.md [hunk 4]
```diff
@@ -1,3 +1,102 @@
-# rukpak
+# RukPak
 
-Rukpak runs in a Kubernetes cluster and defines an API for installing cloud native bundle content 
+RukPak runs in a Kubernetes cluster and defines an API for installing cloud native bundle content.
```
Comment (timflannagan) [inline README.md:3]: Any value in calling out the stage of this project (e.g. pre-alpha) and is subject to change?

File: README.md [hunk 5]
```diff
@@ -1,3 +1,102 @@
...
+      - mountPath: /content
+        configMap:
+          name: resolved-654adh-content
+          namespace: olm
+```
+
+### Provisioner/ProvisionerClass
+`ProvisionerClass` defines a configuration for a `Provisioner`. Provisioners are controllers that understand `Instance`, `Bundle`, and `ProvisionerClass` APIs and take action.
+
+Each `Provisioner` has a unique id. For example, the `Provisioner` that ships with OLM and understand operator bundles has the id: operators.coreos.com/operators.
```
Comment (timflannagan) [inline README.md:81]: I think we'd like to substitute the OLM language here in favor of a rukpak-agnostic provisioner ID: https://github.com/operator-framework/rukpak/blob/main/pkg/k8s/provisioner/bundle_controller.go#L28.

Comment (timflannagan) [inline README.md:81]:
```suggestion
Each `Provisioner` has a unique id. For example, the `Provisioner` that ships with OLM and understands operator bundles has the id: operators.coreos.com/operators.
```

File: README.md (hunk 3 above)
Comment (timflannagan) [reply README.md:16]: Might be worth linking the API type (`./api`) definition directory here.

File: README.md (hunk 4 above)
Comment (timflannagan) [reply README.md:3]: Let's handle this kind of work in a follow-up/further down the line.

File: README.md [hunk 6]
```diff
@@ -1,3 +1,102 @@
...
+spec:
+  class: olm.bundle
+  refs:
+  - file://content
+  volumeMounts:
+  - mountPath: /content
+    configMap:
+      name: local 
+      namespace: plumbus
+  URI: my-bundle.svc.local/content
```
Comment (timflannagan) [inline README.md:43]: This field will likely be exposed outside of the spec, and likely as a field that gets populated post-creation in the status field.

File: README.md [hunk 7]
```diff
@@ -1,3 +1,102 @@
...
+```
+
+### Instance 
+The `Instance` API points to a Bundle and indicates that it should be “active”. This includes pivoting from older versions of an active bundle. 
+`Instance` may also include an embedded spec for a desired Bundle.
+
+Much like Pods stamp out instances of container images, `Instances` stamp out an instance of Bundles. `Instance` can be seen as a generalization of the Pod concept.
+
+The specifics of how an `Instance` makes changes to a cluster based on a referenced `Bundle` is defined by the 
+`ProvisionerClass` and the `Provisioner` that is configured to handle that `ProvisionerClass`.
```
Comment (timflannagan) [inline README.md:53]: @njhale Should we remove an mention of the Instance API until we flesh that API out more and document once that's been implemented, or just document it now as what we perceive as a future API?

**Review by timflannagan — COMMENTED** [review]
We'll need to rebase this PR after I updated the .gitignore field in a previous PR.

I think it's fine to keep the majority of these changes and iterate as we build out this project further. I'd like to strip away any OLM-specific mentions, and just focus on rukpak here. 

I had a couple of comments that I'd like to see addressed (and/or get clarification on) before we merge this, but thanks for taking a stab at this task.

File: README.md (hunk 3 above)
Comment (timflannagan) [reply README.md:16]: We can ignore linking the API directory for now - but I'd like to remove the mention of `Provisioner/ProvisionerClass` and just have `ProvisionerClass` documented here before merging this.

File: README.md (hunk 6 above)
Comment (exdx) [reply README.md:43]: Yes that's correct, it shouldn't be on the spec

**Review by timflannagan — APPROVED** [review]

**Review by njhale — APPROVED** [review]
This looks g2g for now

In the future, I'd like:
- less buzzwords; e.g. "cloud-native"
- more focus on the fact that `rukpak` defines an interface and not an implementation

/lgtm

**Pull request merged**